scripts/fetchai_netutils/fetch/testing/testcase.py

```diff
@@ -0,0 +1,477 @@
+                    node.stop()
+                print('Stopping Node {}...complete'.format(n))
+
+        if self._watchdog:
+            self._watchdog.stop()
+
+    # If something goes wrong, print out debug state (mainly node log files)
+    def dump_debug(self, only_node=None):
+        if self._nodes:
+            for n, node in enumerate(self._nodes):
+
+                if only_node is not None and n is not only_node:
+                    continue
+
+                print('\nNode debug. Node:{}'.format(n))
+                node_log_path = node.log_path
+
+                if not os.path.isfile(node_log_path):
+                    output("Couldn't find supposed node log file: {}".format(
+                        node_log_path))
+                else:
+                    # Send raw bytes directly to stdout since it contains
+                    # non-ascii
+                    data = Path(node_log_path).read_bytes()
+                    sys.stdout.buffer.write(data)
+                    sys.stdout.flush()
+
+
+class DmlfEtchTestCase(TestCase):
+    """
+    Sets up an instance of a test, containing references to started nodes and other relevant data
+    """
+
+    def __init__(self, build_directory, node_exe, yaml_file):
+
+        self._number_of_nodes = 0
+        self._node_connections = None
+        self._port_start_range = 8000
+        self._port_range = 20
+        self._workspace = ""
+        self._max_test_time = 1000
+        self._nodes = []
+        self._watchdog = None
+        self._creation_time = time.perf_counter()
+        self._nodes_pubkeys = []
+        self._nodes_keys = []
+
+        # Default to removing old tests
+        for f in glob.glob(build_directory + "/end_to_end_test_*"):
+            shutil.rmtree(f)
+
+        # To avoid possible collisions, prepend output files with the date
+        self._random_identifer = '{0:%Y_%m_%d_%H_%M_%S}'.format(
+            datetime.datetime.now())
+
+        self._random_identifer = "default"
+
+        self._workspace = os.path.join(
+            build_directory, 'end_to_end_test_{}'.format(
+                self._random_identifer))
+        self._build_directory = build_directory
+        self._node_exe = os.path.abspath(node_exe)
+        self._yaml_file = os.path.abspath(yaml_file)
+        self._test_files_dir = os.path.dirname(self._yaml_file)
+
+        verify_file(node_exe)
+        verify_file(self._yaml_file)
+
+        # Ensure that build/end_to_end_output_XXX/ exists for the test output
+        os.makedirs(self._workspace, exist_ok=True)
+
+    def append_node(self, index, load_directory=None):
+        # Create a folder for the node to write logs to etc.
+        root = os.path.abspath(os.path.join(
+            self._workspace, 'node{}'.format(index)))
+
+        # ensure the workspace folder exits
+        os.makedirs(root, exist_ok=True)
+
+        if load_directory and index in load_directory:
+            load_from = self._test_files_dir + \
```

Comment:
I think it is preffered to use the path utils rather than string concatenation

Reply:
missed that, copy pasted code.
saved from next pr